### PR TITLE
Juan Torres moviesui 1

### PR DESCRIPTION
**Pull request by juandatorr1999**
#### What does this PR do?
Show the list of names and score of the movies separated by categories

#### Where should the reviewer start?
In the MainViewController and in the Components folder where the needed cells were created.

#### Code review checklist [Code Review Rubric](https://github.com/wizeline/iOS-apprenticeship-program-2020/blob/main/code-review-rubric.md)

1. DRY Principle (Don’t repeat yourself)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

2. Yagni Principle (You Aren't Gonna Need It)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

3. KISS Principle (Keep It Simple, St...d)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

4. Remove unused code

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

5. Hard-coded values

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

6. Language style guide & coding conventions
 
 - [Wizeline Style Guide](https://github.com/wizeline/mobile-documentation/blob/master/guidelines/swift-style-guide.md) 
 - [RayWenderlich Style Guide](https://github.com/raywenderlich/swift-style-guide)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

7. Architecture / Design Pattern

    - [ ] MVC  
    - [ ] MVP 
    - [ ] MVVM 
    - [ ] VIPER 

8. Architecture Implementation

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

9. SOLID

- Single Responsibility Principle 

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

- Open Closed Principle

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

- Liskov substitutability principle

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

- Interface segregation

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

- Dependency Inversion principle

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

10. Testing

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

11. Force unwrapping !

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

12. Mutable vs immutable property (Right use of `let`  and `var`)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

13. Retain cycles (Capture list)

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

14. Code Documentation

    - [ ] Excellent    
    - [ ] Good    
    - [ ] Adequate
    - [ ] Developing

### CODE REVIEW

File: TheMovieDb/Components/MainView/MainViewController.swift [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import UIKit
+
+final class MainViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
+  
+  @IBOutlet weak var tableView: UITableView!
+  
+  private var trendingMovies: [Movie] = []
+  
+  // All categories
+  var circularProgressBarView: CircularProgressBarView!
```
Comment (Jerriodas) [inline TheMovieDb/Components/MainView/MainViewController.swift:17]: don't use force unwrap

File: TheMovieDb/Components/MainView/MainViewController.swift [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  }
+    
+  func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+    return CategoriesText.allCases.count
+  }
+  
+  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+    let cell = self.tableView.dequeueReusableCell(withIdentifier: CategoryTableViewCell.identifier) as? CategoryTableViewCell
+    cell?.configure(categoryTitle: CategoriesText.allCases[indexPath.row].rawValue, categories: Categories.allCases[indexPath.row])
+    return cell!
```
Comment (Jerriodas) [inline TheMovieDb/Components/MainView/MainViewController.swift:71]: don't use force unwrap

File: TheMovieDb/Components/Movies Section/CategoryTableViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+//
+//  Created by Juan David Torres on 29/10/21.
+//
+
+import UIKit
+
+final class CategoryTableViewCell: UITableViewCell, UICollectionViewDelegate, UICollectionViewDataSource {
+  
+  @IBOutlet weak var categoryTitle: UILabel!
+  @IBOutlet weak var collectionView: UICollectionView!
```
Comment (Jerriodas) [inline TheMovieDb/Components/Movies Section/CategoryTableViewCell.swift:13]: don't use force unwrap

File: TheMovieDb/Components/Movies Section/CategoryTableViewCell.swift [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    API.getUpcomingMovies.resume(completion: completion)
+  }
+  
+  private func requestAPI() {
+    requestAPITrending()
+    requestAPIPopular()
+    requestAPIUpcoming()
+    requestAPITopRated()
+    requestAPINowPlaying()
+  }
```
Comment (Jerriodas) [inline TheMovieDb/Components/Movies Section/CategoryTableViewCell.swift:124]: this cell implement a lot of logic that could be moved, use Single Responsibility Principle

File: TheMovieDb/Components/Movies Section/MovieCollectionViewCell.swift [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+import UIKit
+
+final class MovieCollectionViewCell: UICollectionViewCell {
+  
+  @IBOutlet weak var movieImage: UIImageView!
+  @IBOutlet weak var movieTitle: UILabel!
+  @IBOutlet weak var movieScore: UILabel!
+  
+  var circularProgressBarView: CircularProgressBarView!
```
Comment (Jerriodas) [inline TheMovieDb/Components/Movies Section/MovieCollectionViewCell.swift:16]: don't use force unwrap

Comment (Jerriodas) [inline TheMovieDb/Components/Movies Section/MovieCollectionViewCell.swift:16]: make it private

File: TheMovieDb/Components/Movies Section/MovieCollectionViewCell.swift [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  var circularViewDuration: TimeInterval = 2
```
Comment (Jerriodas) [inline TheMovieDb/Components/Movies Section/MovieCollectionViewCell.swift:17]: make it private

File: TheMovieDb/Components/Movies Section/MovieCollectionViewCell.swift [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    movieImage.clipsToBounds = true
+    // Movie Title
+    movieTitle.textColor = .black
+    movieTitle.font = UIFont.movieTitle
+    
+    // Movie Title View
+//    movieTitleView.backgroundColor = UIColor.movieTitleViewGrey
+//    movieTitleView.clipsToBounds = true
+//    movieTitleView.layer.cornerRadius = 2
+
```
Comment (Jerriodas) [inline TheMovieDb/Components/Movies Section/MovieCollectionViewCell.swift:45]: If we are not using his can we delete it?

File: TheMovieDb/Components/CircularProgressBarView.swift [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  
+  override init(frame: CGRect) {
+    super.init(frame: frame)
+  }
+  
+  required init?(coder: NSCoder) {
+    super.init(coder: coder)
+  }
+  
+  func createCircularPath() {
```
Comment (Jerriodas) [inline TheMovieDb/Components/CircularProgressBarView.swift:30]: Couldn't see where are we using this func. if we are not using outside the class, we can make it private or remove it if is not used

**Comment by juandatorr1999** [comment]
All corrected, the Requester class has the api requests

File: TheMovieDb/Components/Movies Section/CategoryTableViewCell.swift [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+//
+//  Created by Juan David Torres on 29/10/21.
+//
+
+import UIKit
+
+final class CategoryTableViewCell: UITableViewCell, UICollectionViewDelegate, UICollectionViewDataSource {
+  
+  @IBOutlet weak var categoryLabel: UILabel?
+  @IBOutlet weak var collectionView: UICollectionView?
```
Comment (Jerriodas) [inline TheMovieDb/Components/Movies Section/CategoryTableViewCell.swift:13]: make it private

File: TheMovieDb/Components/Movies Section/MovieCollectionViewCell.swift [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+//
+
+import UIKit
+import Kingfisher
+
+final class MovieCollectionViewCell: UICollectionViewCell {
+  
+  @IBOutlet weak var movieImageView: UIImageView?
+  @IBOutlet weak var movieLabel: UILabel?
+  @IBOutlet weak var scoreLabel: UILabel?
```
Comment (Jerriodas) [inline TheMovieDb/Components/Movies Section/MovieCollectionViewCell.swift:15]: make outlets privates

File: TheMovieDb/Components/Movies Section/Requester.swift [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+          print(response)
+          self?.group.leave()
+          
+        default:
+          self?.movies[.upcomingMovies] = []
+          self?.group.leave()
+        }
+    }
+    API.getUpcomingMovies.resume(completion: completion)
+  }
```
Comment (Jerriodas) [inline TheMovieDb/Components/Movies Section/Requester.swift:117]: use DRY, the code inside each func are doing the same

**Review by Jerriodas — COMMENTED** [review]
just resolve the remaining coments, you're going well

**Review by Jerriodas — APPROVED** [review]

**Pull request merged**